purchase_work_acceptance/models/account_invoice.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2019 Ecosoft Co., Ltd. (http://ecosoft.co.th)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    require_wa = fields.Boolean(
+        compute='_compute_require_wa',
+    )
+    wa_id = fields.Many2one(
+        comodel_name='work.acceptance',
+        string='WA Reference',
+        copy=False,
+        domain=lambda self: [
+            ('state', '=', 'accept'),
+            ('purchase_id', '=', self._context.get('active_id'))],
+        help='To control quantity and unit price of the vendor bill, to be '
+             'according to the quantity and unit price of the work acceptance.',
+    )
+
+    @api.multi
+    def _compute_require_wa(self):
+        self.require_wa = self.env.user.has_group(
+            'purchase_work_acceptance.group_work_acceptance_enforce')
+
+    def _prepare_invoice_line_from_po_line(self, line):
+        res = super()._prepare_invoice_line_from_po_line(line)
+        wa_line = self.wa_id.wa_line_ids.filtered(
+            lambda l: l.purchase_line_id == line)
+        if wa_line:
+            res['quantity'] = wa_line['product_qty']
```

Comment:
I will change `wa_line['product_qty']` to `wa_line.product_qty`.

But I don't understand why I should use the minimum between `wa_line.product_qty` and `res['quantity']`. I need to quantity of invoice = quantity of wa when choose wa.

Reply:
If you create a PO with one line and create a WA with this line you can create two invoices for this product. This will give problems, obviously, as you will create two invoices paying the same WA/PO.
I have already tested on the runbot:

- Enable WA (forcing WA for invoice)
- Create a PO with one line with the product as a service
- Confirm PO
- Create a WA and accept WA
- Create invoice with this WA and confirm it
- Create another invoice with this WA and confirm it

You can check the PO00010 of the current runbot to check it

Reply:
Now, I don't allow create many invoices same WA. It should 1 WA : 1 Invoice